File: 01-week-1--starter-algorithms/01-day-2--find-first-duplicate/javascript/find_first_duplicate.js
function findFirstDuplicate(arr) {
  // type your code here
  const uniques = new Set();

  for (const value of arr) {
    if (uniques.has(value)) {
      return value;
    }

    uniques.add(value);
  }

  return -1;
}
//   let myHash= {}
//   let firstDupe;
//   for(let i=0;i<arr.length;i++){
//     myHash[arr[i]] ? myHash[arr[i]] = myHash[arr[i]]+1 : myHash[arr[i]] = 1
//       //firstDupe = arr[i]
//     // } else {
//     //   myHash[arr[i]] = 1
//     // }
//   }
//   let vals = Object.values(myHash)
//   let keys = Object.keys(myHash)
//    firstDupe = vals.find((el, ind) => el > 1 ? ind : -1)
//    //(ar=> myHash[ar]>1)
   
//    let realDupe =  keys[firstDupe] 
//   console.log(myHash, realDupe, firstDupe)
// }
//   arr.map(element => {
//     myHash[element] ? firstDupe = element : myHash[element] = 1;
    
//   });
//   if(firstDupe){
//     return firstDupe
//   } else
//   return -1
// }

console.log(findFirstDuplicate([2, 1, 3, 3, 2]))
console.log(findFirstDuplicate([1,2,3,4]))

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 3");
  console.log("=>", findFirstDuplicate([2, 1, 3, 3, 2]));

  console.log("");

  console.log("Expecting: -1");
  console.log("=>", findFirstDuplicate([1, 2, 3, 4]));
}

module.exports = findFirstDuplicate;

// Please add your pseudocode to this file
// And a written explanation of your solution

File: 01-week-1--starter-algorithms/01-day-2--find-first-duplicate/ruby/find_first_duplicate.rb
def find_first_duplicate(arr)
  # type your code in here
  #setting up second array to push values into if not already in
  garb= []
  i=0
    arr.each do |el|
      garb << el unless garb.include?(el)
    end
    
    while i < arr.length do
     if arr[i] != garb[i]
      then return arr[i]
     
  end
   i+=1
 end
 -1
end

# HEY THIS IS THE SOLUTION'S FILE WAY
# def find_first_duplicate(arr)
#   uniques = Set.new

#   arr.each do |value|
#     return value if uniques.include?(value)

#     uniques.add(value)
#   end

#   -1
# end



# find_first_duplicate([1,2,3,4])
# puts find_first_duplicate([1,2,3,3,2])



if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 3"
  puts "=>", find_first_duplicate([2, 1, 3, 3, 2])

  puts

  puts "Expecting: -1"
  puts "=>", find_first_duplicate([1, 2, 3, 4])

  # Don't forget to add your own!
end



# Please add your pseudocode to this file
# And a written explanation of your solution
